homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,257 @@
+        )
+        return
+
+    manager.skip_update(msg["domain"], msg["identifier"], msg["version"])
+    connection.send_result(msg["id"])
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "update/update",
+        vol.Required("domain"): str,
+        vol.Required("identifier"): str,
+        vol.Required("version"): str,
+        vol.Optional("backup"): bool,
+    }
+)
+@websocket_api.async_response
+async def handle_update(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Handle an update."""
+    manager: UpdateManager = hass.data[DOMAIN]
+
+    if not manager.domain_is_valid(msg["domain"]):
+        connection.send_error(
+            msg["id"], websocket_api.ERR_NOT_FOUND, "Domain not supported"
+        )
+        return
+
+    if not await manager.perform_update(
+        domain=msg["domain"],
+        identifier=msg["identifier"],
+        version=msg["version"],
+        backup=msg.get("backup"),
+    ):
+        connection.send_error(msg["id"], "update_failed", "Update failed")
+        return
+
+    connection.send_result(msg["id"])
+
+
+class UpdatePlatformProtocol(Protocol):
+    """Define the format that update platforms can have."""
+
+    async def async_list_updates(self, hass: HomeAssistant) -> list[UpdateDescription]:
+        """List all updates available in the integration."""
+
+    async def async_perform_update(
+        self,
+        hass: HomeAssistant,
+        identifier: str,
+        version: str,
+        **kwargs: Any,
+    ) -> bool:
+        """Perform an update."""
+
+
+@dataclasses.dataclass()
+class UpdateDescription:
+    """Describe an update update."""
+
+    identifier: str
+    name: str
+    current_version: str
+    available_version: str
+    changelog_content: str | None = None
+    changelog_url: str | None = None
+    icon_url: str | None = None
+    provides_backup: bool = False
+
+
+class UpdateManager:
+    """Update manager for the update integration."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the update manager."""
+        self._hass = hass
+        self._store = storage.Store(
+            hass=hass,
+            version=STORAGE_VERSION,
+            key=DOMAIN,
+        )
+        self._skip: set[str] = set()
+        self._platforms: dict[str, UpdatePlatformProtocol] = {}
+        self._loaded = False
+
+    def add_platform(
+        self,
+        integration_domain: str,
+        platform: UpdatePlatformProtocol,
+    ) -> None:
+        """Add a platform to the update manager."""
+        self._platforms[integration_domain] = platform
+
+    async def gather_updates(self) -> list[dict[str, Any]]:
+        """Gather updates."""
+        if not self._loaded:
+            await integration_platform.async_process_integration_platforms(
+                self._hass, DOMAIN, _register_update_platform
```

Comment:
Can we use `self.add_platform` directly now ?

Reply:
no

Reply:
It would work if `add_platform` becomes async?

Reply:
How? the signature does not match? `self: UpdateManager` / `hass: HomeAssistant`

Reply:
Then make the signature match. It feels weird to have a class have an external method to register a callback to call a public method to register a platform. That can just be all internal.

Reply:
My bad... it passes `self.add_platform`, this works :+1: